src/pybind/mgr/dashboard/controllers/__init__.py

```diff
@@ -25,23 +20,53 @@
 from ..services.exception import serialize_dashboard_exception
 
 
-def ApiController(path):
-    def decorate(cls):
+class Controller(object):
+    def __init__(self, path, base_url=None):
+        self.path = path
+        self.base_url = base_url
+
+        if self.path and self.path[0] != "/":
+            self.path = "/" + self.path
+
+        if self.base_url is None:
+            self.base_url = ""
+        elif self.base_url == "/":
+            self.base_url = ""
+
+        if self.base_url == "" and self.path == "":
+            self.base_url = "/"
+
+    def __call__(self, cls):
         cls._cp_controller_ = True
-        cls._cp_path_ = path
+        cls._cp_path_ = "{}{}".format(self.base_url, self.path)
+
         config = {
             'tools.sessions.on': True,
             'tools.sessions.name': Session.NAME,
             'tools.session_expire_at_browser_close.on': True,
-            'tools.dashboard_exception_handler.on': True,
+            'tools.dashboard_exception_handler.on': True
         }
         if not hasattr(cls, '_cp_config'):
             cls._cp_config = {}
         if 'tools.authenticate.on' not in cls._cp_config:
             config['tools.authenticate.on'] = False
         cls._cp_config.update(config)
         return cls
-    return decorate
+
+
+class ApiController(Controller):
+    def __init__(self, path, version=1):
+        if version == 1:
+            base_url = "/api"
+        else:
+            base_url = "/api/v" + str(version)
```

Comment:
Is there a valid reason for adding an API version right now? Adding this seems strange to me, especially as there are better alternatives than adding a version to the URL path.

Reply:
With the current way the HTTP requests are dispatched, only the URL based versioning is possible. To support alternative approaches like passing the version in the request headers, we require a different dispatching implementation.

Reply:
I see this as an investment in the future. While the current API will still be in flux and evolving, we might need to "freeze" it at some point as a reference for existing applications that might depend on a certain behaviour, while the dashboard UI may require different API calls in order to further evolve. For the time being, the URL path should not be "polluted" by a version element, but we should reserve the option of enabling it if needed (even though I too would be more in favor of including the version information in the request headers instead)

Reply:
> there are better alternatives  than adding a version to the URL path

@sebastian-philipp out of curiosity, what alternative are you referring to?

Reply:
There are so many downsides related to adding a version schema to the path.

1. It will not really help us in providing a stable API as long as we don't have a complete solution for dealing with underlying C++-API changes. We will just fool ourselves and our users in thinking that the API is stable, if it is versioned. 

2. When do you want to switch to a new global API version? Typically, if you're re-implementing the backend and want to replace the backend with something different. But then, there is no need for implementing this now. 

3. Or do you want to change the API version for specific calls? Every time you're doing an increment like this, you're making the API much more complicated to use and to maintain. We really should avoid that, if possible. Also then, the version number would too early in the URL path. 

4. We should keep the URL paths as human friendly as possible, as they are the ones using our naming scheme. A version here would be a strange mix of a documentation thingy and a technical requirement, which serves no real useful purpose. 

5. If at some point in the future, we're facing with an issue that needs to be solved with versioning, we can add HTTP-header based dispatching at any time. There is no need to implement this now.

6. And finally, instead of adding a versioning scheme, we should instead focus of a nicely designed API that's easy to use, easy to understand and easy to enhance. Non-breaking API changes are always the best way forward.

Reply:
@sebastian-philipp most of the points you mentioned are applicable to any kind of API versioning scheme and not just for URL based schemes.

But I agree with you, API versioning is not needed now and it's not as easy as just start implementing new versions of our controllers and by magic the previous versions remain stable behavior/semantic wise.

That said, I'll remove the versioning part as it's just a few lines and nothing in the remaining code depends on that.

Reply:
@sebastian-philipp thanks for the clarification 

@rjfd I agree, ATM we don’t use api version so we can discuss and implement this later